integration/gossip/gossip_test.go

```diff
@@ -140,6 +142,167 @@ var _ = Describe("Gossip State Transfer", func() {
+
+			By("restarting org2 peers and wait for reconnection")
+			startPeers(nwprocs, false, peer0Org2, peer1Org2)
+			time.Sleep(2 * core.Peer.Gossip.ReconnectInterval)
+
+			By("verifying membership on peer1Org1, expected to discover peers in a different org")
+			Eventually(nwo.DiscoverPeers(network, peer1Org1, "User1", "testchannel"), network.EventuallyTimeout).Should(ConsistOf(
+				network.DiscoveredPeer(peer1Org1, "_lifecycle"),
+				network.DiscoveredPeer(peer0Org2, "_lifecycle"),
+				network.DiscoveredPeer(peer1Org2, "_lifecycle"),
+			))
+		})
+
+		It("updates membership when non-anchor peer in the same org is stopped/restarted", func() {
+			peer0Org1 := network.Peer("Org1", "peer0")
+			peer1Org1 := network.Peer("Org1", "peer1")
+
+			By("bringing up all peers")
+			startPeers(nwprocs, false, peer0Org1, peer1Org1)
+
+			By("creating and joining a channel")
+			network.CreateChannel(channelName, orderer, peer0Org1)
+			network.JoinChannel(channelName, orderer, peer0Org1, peer1Org1)
+			network.UpdateChannelAnchors(orderer, channelName)
+
+			By("verifying membership on anchor peer peer0Org1")
+			Eventually(nwo.DiscoverPeers(network, peer0Org1, "User1", "testchannel"), network.EventuallyTimeout).Should(ConsistOf(
+				network.DiscoveredPeer(peer0Org1, "_lifecycle"),
+				network.DiscoveredPeer(peer1Org1, "_lifecycle"),
+			))
+
+			By("stopping non-anchor peer peer1Org1 to simulate lost connections")
+			stopPeers(nwprocs, peer1Org1)
+
+			// Sleep the same amount of time as it takes to remove a message from the aliveMsgStore, and add a second as buffer
+			core := network.ReadPeerConfig(peer0Org1)
+			time.Sleep(core.Peer.Gossip.AliveExpirationTimeout*time.Duration(core.Peer.Gossip.MsgExpirationFactor) + time.Second)
+
+			By("verifying membership on peer0Org1 after all other peers are stopped")
+			Eventually(nwo.DiscoverPeers(network, peer0Org1, "User1", "testchannel"), network.EventuallyTimeout).Should(ConsistOf(
+				network.DiscoveredPeer(peer0Org1, "_lifecycle"),
+			))
+			runner := nwprocs.peerRunners[peer0Org1.ID()]
+			Eventually(runner.Err(), network.EventuallyTimeout, time.Second).Should(gbytes.Say("Removing member: Endpoint: %s", peerEndpoints[peer1Org1.ID()]))
+
+			By("restarting Org1 peers and wait for reconnection")
+			startPeers(nwprocs, false, peer1Org1)
+			time.Sleep(2 * core.Peer.Gossip.ReconnectInterval)
+
+			By("verifying membership on peer0Org1, expected to discover peers in the same org")
+			Eventually(nwo.DiscoverPeers(network, peer0Org1, "User1", "testchannel"), network.EventuallyTimeout).Should(ConsistOf(
+				network.DiscoveredPeer(peer0Org1, "_lifecycle"),
+				network.DiscoveredPeer(peer1Org1, "_lifecycle"),
+			))
+		})
+
+		// TODO: it causes test flake, where the non-anchor peer cannot connect to the anchor peer after the anchor peer is stop/restarted
+		It("updates membership when anchor peer in the same org is stopped/restarted", func() {
```

Comment:
@yacovm I moved IT test from discovery to gossip. This test still fails on my laptop.

Reply:
This is because it tries to connect to the internal endpoint instead of the external endpoint, but the internal endpoint is not a loopback address, and the default core_template is designed to bind only to the loopback and not to 0.0.0.0. 
`[e][Org1.peer1] 2020-06-23 14:06:56.592 IDT [gossip.discovery] func1 -> DEBU 099 Endpoint: 127.0.0.1:21504, InternalEndpoint: 192.168.1.21:21504, PKI-ID: ff2b9769feb64436a78c030d0d43cb051b46297850d5d1edddbf8d57a5fd400b, Metadata:  is still dead
`

The reason this works with cross organization peers, is that cross org communication use the external endpoint which is configured to use the loopback address, so it acts as a proxy between peers of the same org. 

What we need is to simply force the peers use internal endpoints with a loopback address. 


One way of solving this, is to change the core template to have:  `endpoint: 127.0.0.1:{{ .PeerPort Peer "Listen" }}` in the gossip section of the core template.

Reply:
@yacovm Thank you. I have updated the code to overwrite `addressAutoDetect` to false in gossip_test.go and it fixed the problem. Regarding to not setting `addressAutoDetect` in the core template, will use https://jira.hyperledger.org/browse/FAB-18018.